plutus-chain-index-core/src/Marconi/Index/ScriptTx.hs

```diff
@@ -0,0 +1,194 @@
+import Cardano.Ledger.Shelley.Scripts qualified as LedgerShelley
+import Cardano.Ledger.ShelleyMA.Timelocks qualified as Timelock
+
+import RewindableIndex.Index.VSqlite (SqliteIndex)
+import RewindableIndex.Index.VSqlite qualified as Ix
+
+
+newtype Depth = Depth Int
+
+newtype ScriptAddress = ScriptAddress Shelley.ScriptHash
+  deriving (Show, Eq)
+newtype TxCbor = TxCbor BS.ByteString
+  deriving (Show)
+  deriving newtype (SQL.ToField, SQL.FromField)
+
+-- * SQLite
+
+data ScriptTxRow = ScriptTxRow
+  { scriptAddress :: !ScriptAddress
+  , txCbor        :: !TxCbor
+  } deriving (Generic)
+
+instance SQL.ToField ScriptAddress where
+  toField (ScriptAddress hash)  = SQL.SQLBlob . Shelley.serialiseToRawBytes $ hash
+instance SQL.FromField ScriptAddress where
+  fromField f = SQL.fromField f >>=
+    either
+      (const cantDeserialise)
+      (\b -> maybe cantDeserialise (return . ScriptAddress) $ Shelley.deserialiseFromRawBytes Shelley.AsScriptHash b) . deserialiseOrFail
+    where
+      cantDeserialise = SQL.returnError SQL.ConversionFailed f "Cannot deserialise address."
+
+instance SQL.ToRow ScriptTxRow where
+  toRow o = [SQL.toField $ scriptAddress o, SQL.toField $ txCbor o]
+
+-- * Indexer
+
+type Query = ScriptAddress
+type Result = [TxCbor]
+
+data ScriptTxUpdate = ScriptTxUpdate
+  { txScripts :: [(TxCbor, [ScriptAddress])]
+  , slotNo    :: !SlotNo
+  } deriving (Show)
+
+type ScriptTxIndex = SqliteIndex ScriptTxUpdate () Query Result
+
+
+toUpdate :: forall era . C.IsCardanoEra era => [C.Tx era] -> SlotNo -> ScriptTxUpdate
+toUpdate txs = ScriptTxUpdate txScripts'
+  where
+    txScripts' = map (\tx -> (TxCbor $ C.serialiseToCBOR tx, getTxScripts tx)) txs
+
+getTxBodyScripts :: forall era . C.TxBody era -> [ScriptAddress]
+getTxBodyScripts body = let
+    hashesMaybe :: [Maybe C.ScriptHash]
+    hashesMaybe = case body of
+      Shelley.ShelleyTxBody shelleyBasedEra _ scripts _ _ _ -> flip map scripts $ \script ->
+        case fromShelleyBasedScript shelleyBasedEra script of
+          Shelley.ScriptInEra _ script' -> Just $ C.hashScript script'
+      _ -> [] -- Byron transactions have no scripts
+    hashes = catMaybes hashesMaybe :: [Shelley.ScriptHash]
+  in map ScriptAddress hashes
+
+getTxScripts :: forall era . C.Tx era -> [ScriptAddress]
+getTxScripts (C.Tx txBody _ws) = getTxBodyScripts txBody
+
+open :: FilePath -> Depth -> IO ScriptTxIndex
+open dbPath (Depth k) = do
+  ix <- fromJust <$> Ix.newBoxed query store onInsert k ((k + 1) * 2) dbPath
+  let c = ix ^. Ix.handle
+  SQL.execute_ c "CREATE TABLE IF NOT EXISTS script_transactions (scriptAddress TEXT NOT NULL, txCbor BLOB NOT NULL)"
+  SQL.execute_ c "CREATE INDEX IF NOT EXISTS script_address ON script_transactions (scriptAddress)"
+  pure ix
+
+  where
+    onInsert :: ScriptTxIndex -> ScriptTxUpdate -> IO [()]
+    onInsert _ix _update = pure []
+
+    store :: ScriptTxIndex -> IO ()
+    store ix = do
+      persisted <- Ix.getEvents $ ix ^. Ix.storage
```

Comment:
We should not store the events, only what we have in the buffer.

Reply:
I now only store the buffered events here 83ac994f4f7b457f6d5517adca3d590a070fbc09 and think there was the same bug in the Utxo indexer as well, does it look ok now there as well?